### PR TITLE
Modernize startup error pages

### PR DESCRIPTION
**Pull request by pakrym**
image

image

image

### CODE REVIEW

**Comment by pakrym** [comment]
We still need to create different fwlinks for all pages.

**Comment by Tratcher** [comment]
What's the context for these pages? The first two look identical.

**Comment by Tratcher** [comment]
Needs bullets.

**Comment by shirhatti** [comment]
Existing out-of-proc fwlink 808681
New in-proc fwlinks 2028265, 2028526

**Comment by jkotalik** [comment]
For #1470

File: src/AspNetCoreModuleV2/AspNetCore/InProcessShimStaticHtml.htm [hunk 1]
```diff
@@ -1,45 +1,83 @@
...
+
+        </style>
+    </head>
+    <body>
+        <h1> HTTP Error 500.0 - ANCM In-Process Startup Failure </h1>
+
+        <h2> Common causes of this issue: </h2>
+
+        <ul>
+            <li> The specified version of Microsoft.NetCore.App or Microsoft.AspNetCore.App was not found. </li>
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/AspNetCore/InProcessShimStaticHtml.htm]: @shirhatti please review the list of recommendations here (for language/effectiveness)

**Comment by pakrym** [comment]
Added fwlinks, bullets and support for `ANCM_ADDITIONAL_ERROR_PAGE_TEXT` environment variable to add custom text to error pages.

File: src/AspNetCoreModuleV2/AspNetCore/InProcessShimStaticHtml.htm [hunk 2]
```diff
@@ -1,45 +1,85 @@
...
-    </div>
-</body>
+<!DOCTYPE html>
+<html lang="en-US" xmlns="http://www.w3.org/1999/xhtml">
+    <head>
+        <meta charset="utf-8" />
+        <title> HTTP Error 500.0 - ANCM In-Process Startup Failure </title>
+        <style>
+            body {
+                font-family: 'Segoe UI', Tahoma, Arial, Helvetica, sans-serif;
```
Comment (shirhatti) [inline src/AspNetCoreModuleV2/AspNetCore/InProcessShimStaticHtml.htm]: I'm not sure how licensing of these fonts work. We'll need to verify if this is kosher

File: src/AspNetCoreModuleV2/AspNetCore/OutOfProcessShimStaticHtml.htm [hunk 3]
```diff
@@ -76,5 +76,7 @@ <h2> Troubleshooting steps: </h2>
         <h2> For more information visit:
             <a href="https://go.microsoft.com/fwlink/?LinkID=808681"> <cite> https://go.microsoft.com/fwlink/?LinkID=808681 </cite> </a>
         </h2>
+
+        %S
```
Comment (shirhatti) [inline src/AspNetCoreModuleV2/AspNetCore/OutOfProcessShimStaticHtml.htm]: What's this?

Comment (pakrym) [reply src/AspNetCoreModuleV2/AspNetCore/OutOfProcessShimStaticHtml.htm]: Placeholder for `ANCM_ADDITIONAL_ERROR_PAGE_TEXT `

File: src/AspNetCoreModuleV2/AspNetCore/InProcessShimStaticHtml.htm (hunk 2 above)
Comment (pakrym) [reply src/AspNetCoreModuleV2/AspNetCore/InProcessShimStaticHtml.htm]: Copied from https://github.com/aspnet/Diagnostics/blob/release/2.2/src/Microsoft.AspNetCore.Diagnostics/DeveloperExceptionPage/Views/ErrorPage.css

Comment (shirhatti) [reply src/AspNetCoreModuleV2/AspNetCore/InProcessShimStaticHtml.htm]: Works for me

**Comment by jkotalik** [comment]
Are we going to add the Antares specific on here too (hard coded)?

**Comment by pakrym** [comment]
No, we provide a way to add additional content to error pages for them via `ANCM_ADDITIONAL_ERROR_PAGE_TEXT` environment variable

**Comment by pakrym** [comment]
@shirhatti do you prefer `ANCM_ADDITIONAL_ERROR_PAGE_TEXT` to be html or just a link?

**Comment by jkotalik** [comment]
Fix GlobalVersion_NewVersionNumber_Fails as it checks for the out of process error page.

**Comment by pakrym** [comment]
Yep

**Comment by davidfowl** [comment]
Did we also fix the timeout error for out of proc?

**Comment by rockerinthelocker** [comment]
For what are the leading/trailing white-spaces within HTML tags good for? Get they stripped before templates are sent?

**Comment by pakrym** [comment]
Switched env var to be `ANCM_ADDITIONAL_ERROR_PAGE_LINK`:

image

File: src/AspNetCoreModuleV2/OutOfProcessRequestHandler/forwardinghandler.cpp [hunk 4]
```diff
@@ -318,18 +319,8 @@ FORWARDING_HANDLER::OnExecuteRequestHandler()
...
-            "text/html",
-            (USHORT)strlen("text/html"),
-            FALSE
-        );
-
-        DataChunk.DataChunkType = HttpDataChunkFromMemory;
-        DataChunk.FromMemory.pBuffer = (PVOID)sm_pStra502ErrorMsg.QueryStr();
-        DataChunk.FromMemory.BufferLength = sm_pStra502ErrorMsg.QueryCB();
-        pResponse->WriteEntityChunkByReference(&DataChunk);
+        ServerErrorHandler handler(*m_pW3Context, 502, "Bad Gateway", hr, g_hOutOfProcessRHModule, m_pApplication->QueryConfig()->QueryDisableStartUpErrorPage(), OUT_OF_PROCESS_RH_STATIC_HTML);
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/OutOfProcessRequestHandler/forwardinghandler.cpp]: How does this set the substatus code?

File: test/IISExpress.FunctionalTests/InProcess/StartupTests.cs [hunk 5]
```diff
@@ -40,6 +40,50 @@ public async Task ExpandEnvironmentVariableInWebConfig()
...
+
+            Assert.Contains("HTTP Error 500.0 - ANCM In-Process Handler Load Failure", await response.Content.ReadAsStringAsync());
+
+            Assert.Contains("http://example", await response.Content.ReadAsStringAsync());
+        }
+
+        [ConditionalFact]
+        public async Task IncludesAdditionalErrorPageTextOutOfProcess()
+        {
+            var deploymentParameters = _fixture.GetBaseDeploymentParameters(HostingModel.OutOfProcess, publish: true);
```
Comment (jkotalik) [inline test/IISExpress.FunctionalTests/InProcess/StartupTests.cs]: Make a helper between the two tests.

File: src/AspNetCoreModuleV2/AspNetCore/ApplicationFactory.h [hunk 6]
```diff
@@ -31,13 +31,17 @@ class ApplicationFactory
...
     HRESULT Execute(
         _In_  IHttpServer           *pServer,
-        _In_  const IHttpApplication *pHttpApplication,
-        _Outptr_ IAPPLICATION       **pApplication) const noexcept
+        _In_  IHttpContext          *pHttpContext,
+        _Outptr_ IAPPLICATION       **pApplication) const
     {
-        std::array<APPLICATION_PARAMETER, 1> parameters {
-            {"InProcessExeLocation", m_location.data()}
+        std::array<APPLICATION_PARAMETER, 2> parameters {
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/AspNetCore/ApplicationFactory.h]: Remove this.

File: src/AspNetCoreModuleV2/OutOfProcessRequestHandler/forwardinghandler.cpp (hunk 4 above)
Comment (pakrym) [reply src/AspNetCoreModuleV2/OutOfProcessRequestHandler/forwardinghandler.cpp]: Good point, it doesn't.

File: src/AspNetCoreModuleV2/InProcessRequestHandler/StartupExceptionApplication.h [hunk 7]
```diff
@@ -28,7 +28,7 @@ class StartupExceptionApplication : public InProcessApplicationBase
     HRESULT
         CreateHandler(IHttpContext *pHttpContext, IREQUEST_HANDLER ** pRequestHandler)
     {
-        *pRequestHandler = new ServerErrorHandler(*pHttpContext, 500, "Internal Server Error", m_HR, m_moduleInstance, m_disableLogs, IN_PROCESS_RH_STATIC_HTML);
+        *pRequestHandler = new ServerErrorHandler(*pHttpContext, 500, 0, "Internal Server Error", m_HR, m_moduleInstance, m_disableLogs, IN_PROCESS_RH_STATIC_HTML);
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/StartupExceptionApplication.h]: Shouldn't this be 30?

File: src/AspNetCoreModuleV2/OutOfProcessRequestHandler/forwardinghandler.cpp [hunk 8]
```diff
@@ -5,6 +5,7 @@
 #include "url_utility.h"
 #include "exceptions.h"
 #include "ServerErrorHandler.h"
+#include "../AspNetCore/resource.h"
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/OutOfProcessRequestHandler/forwardinghandler.cpp]: You shouldn't need to include anything in AspNetCore into the oop request handler.

File: test/IISExpress.FunctionalTests/InProcess/StartupTests.cs [hunk 9]
```diff
@@ -84,6 +78,18 @@ public async Task IncludesAdditionalErrorPageTextOutOfProcess()
             Assert.Contains("<a href=\"http://example\"> <cite> http://example </cite></a> and ", await response.Content.ReadAsStringAsync());
         }
 
+        private async Task<HttpResponseMessage> DeployAppWithStartupFailure(IISDeploymentParameters deploymentParameters)
```
Comment (jkotalik) [inline test/IISExpress.FunctionalTests/InProcess/StartupTests.cs]: You can share this with IncludesAdditionalErrorPageTextOutOfProcess

**Comment by pakrym** [comment]
> Did we also fix the timeout error for out of proc?

No, that's a separate issue.

**Comment by shirhatti** [comment]
> can we schedule a meeting to go through all language used for ANCM errors/responses?

Ack

**Comment by shirhatti** [comment]
Additional FWLink for out-of-process: 2028526

**Comment by DamianEdwards** [comment]
This looks great! I'd love to see what it looks like when App Service adds the environment-specific content too, e.g. adding a link to where to view logs.

**Pull request merged**